ast/src/parsed/visitor.rs

```diff
@@ -0,0 +1,385 @@
+        }
+    }
+}
+
+impl<T> ExpressionVisitor<T, ShiftedPolynomialReference<T>> for ArrayExpression<T> {
+    fn visit_expressions_mut<F, B>(&mut self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&mut Expression<T>) -> ControlFlow<B>,
+    {
+        match self {
+            ArrayExpression::Value(expressions) | ArrayExpression::RepeatedValue(expressions) => {
+                expressions
+                    .iter_mut()
+                    .try_for_each(|e| e.visit_expressions_mut(f, o))
+            }
+            ArrayExpression::Concat(a1, a2) => [a1, a2]
+                .iter_mut()
+                .try_for_each(|e| e.visit_expressions_mut(f, o)),
+        }
+    }
+
+    fn visit_expressions<F, B>(&self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&Expression<T>) -> ControlFlow<B>,
+    {
+        match self {
+            ArrayExpression::Value(expressions) | ArrayExpression::RepeatedValue(expressions) => {
+                expressions
+                    .iter()
+                    .try_for_each(|e| e.visit_expressions(f, o))
+            }
+            ArrayExpression::Concat(a1, a2) => {
+                [a1, a2].iter().try_for_each(|e| e.visit_expressions(f, o))
+            }
+        }
+    }
+}
+
+impl<T, Ref> ExpressionVisitor<T, Ref> for LambdaExpression<T, Ref> {
+    fn visit_expressions_mut<F, B>(&mut self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&mut Expression<T, Ref>) -> ControlFlow<B>,
+    {
+        self.body.visit_expressions_mut(f, o)
+    }
+
+    fn visit_expressions<F, B>(&self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&Expression<T, Ref>) -> ControlFlow<B>,
+    {
+        self.body.visit_expressions(f, o)
+    }
+}
+
+impl<T, Ref> ExpressionVisitor<T, Ref> for ArrayLiteral<T, Ref> {
+    fn visit_expressions_mut<F, B>(&mut self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&mut Expression<T, Ref>) -> ControlFlow<B>,
+    {
+        self.items
+            .iter_mut()
+            .try_for_each(|item| item.visit_expressions_mut(f, o))
+    }
+
+    fn visit_expressions<F, B>(&self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&Expression<T, Ref>) -> ControlFlow<B>,
+    {
+        self.items
+            .iter()
+            .try_for_each(|item| item.visit_expressions(f, o))
+    }
+}
+
+impl<T, Ref> ExpressionVisitor<T, Ref> for FunctionCall<T, Ref> {
+    fn visit_expressions_mut<F, B>(&mut self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&mut Expression<T, Ref>) -> ControlFlow<B>,
+    {
+        self.arguments
+            .iter_mut()
+            .try_for_each(|item| item.visit_expressions_mut(f, o))
+    }
+
+    fn visit_expressions<F, B>(&self, f: &mut F, o: VisitOrder) -> ControlFlow<B>
+    where
+        F: FnMut(&Expression<T, Ref>) -> ControlFlow<B>,
```

Comment:
If we could somehow shorten the signature here...

Reply:
I really prefer passing lambda functions instead of actual proper traits. I see that this is needed for something like the Folder, but since the signature is always the same for the simple recursive type of `Expression`, we can use lambda functions instead.

Again, this is the trade-off between usability from the outside and how complicated the mechanism is in the inside.

We can probably get away with removing the `B` type altogether (and replace it by `()`), because we can also just pass a return value through a closure instead of through the `ControlFlow` mechanism.

Would you prefer that?